### DISCUSSION TITLE
How to implement a AttributeHistoryServiceSet in OPC UA Server

### DISCUSSION THREAD

**Discussion by zhanglc** [eclipse-milo/milo#1067] [General]
Hi ,

I want to use `milo` to implement an opc ua server , in which I can mock some history data. 

Looked at the milo code and it seems that I need to implement an `AttributeHistoryServiceSet` , and then set in the `OpcUaServer`, when the client call, the subclass of `AttributeHistoryServiceSet` will return data , which is mocked.

How do I do it? Is there some tutorial here? 

Any tips would be greatly appreciated.

**Comment by kevinherron** [comment]
In your `Namespace` or `AddressSpace` implementation you would just implement/override the default implementations of `AttributeHistoryServices::historyRead` and `AttributeHistoryServices::historyUpdate` https://github.com/eclipse/milo/blob/3419235853491ac89f3391bffd337c8200522522/opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/api/services/AttributeHistoryServices.java#L32-L73)

You should not need to replace the service sets.
